core/unit_test/TestViewCtorProp.hpp

```diff
@@ -0,0 +1,145 @@
+
+  static_assert(std::is_constructible_v<vcp_label_base_t, char*>);
+}
+
+// Check traits of derived class Kokkos::Impl::ViewCtorProp<std::string>.
+TEST(TEST_CATEGORY, vcp_label_traits) {
+  using vcp_label_base_t = Kokkos::Impl::ViewCtorProp<void, std::string>;
+  using vcp_label_t      = Kokkos::Impl::ViewCtorProp<std::string>;
+
+  static_assert(std::is_base_of_v<vcp_label_base_t, vcp_label_t>);
+
+  static_assert(vcp_label_t::has_label);
+
+  // Check that the derived class is not default constructible. It is a design
+  // choice to not allow the default constructor to be called.
+  static_assert(!std::is_default_constructible_v<vcp_label_t>);
+
+  static_assert(std::is_constructible_v<vcp_label_t, std::string>);
+  static_assert(std::is_constructible_v<vcp_label_t, const std::string&>);
+  static_assert(std::is_constructible_v<vcp_label_t, std::string&&>);
+
+  constexpr unsigned N = 3;
+  static_assert(std::is_constructible_v<vcp_label_t, const char[N]>);
+  static_assert(std::is_constructible_v<vcp_label_t, char[N]>);
+
+  static_assert(std::is_constructible_v<vcp_label_t, char*>);
+}
+
+// Check that the constructor of Kokkos::Impl::ViewCtorProp<std::string>
+// moves a label passed by rvalue reference.
+TEST(TEST_CATEGORY, vcp_label_constructor_can_move) {
+  using vcp_label_t = Kokkos::Impl::ViewCtorProp<std::string>;
+
+  std::string label("our label");
+
+  vcp_label_t prop(std::move(label));
+
+  ASSERT_TRUE(label.empty());
+  ASSERT_EQ(Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop),
+            "our label");
+}
+
+// Check that Kokkos::view_alloc moves a label passed by rvalue reference.
+TEST(TEST_CATEGORY, vcp_label_view_alloc_can_move) {
+  std::string label("our label");
+
+  auto prop = Kokkos::view_alloc(std::move(label));
+
+  ASSERT_TRUE(label.empty());
+  ASSERT_EQ(Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop),
+            "our label");
+}
+
+// Check the copy constructor of Kokkos::Impl::ViewCtorProp<std::string>.
+TEST(TEST_CATEGORY, vcp_label_copy_constructor) {
+  using vcp_empty_t = Kokkos::Impl::ViewCtorProp<>;
+  using vcp_label_t = Kokkos::Impl::ViewCtorProp<std::string>;
+
+  // Copy construction from an empty view constructor properties object.
+  static_assert(std::is_constructible_v<vcp_label_t, const vcp_empty_t&>);
+
+  vcp_empty_t prop_empty;
+  vcp_label_t prop_empty_copy(prop_empty);
+
+  ASSERT_TRUE(
+      Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop_empty_copy)
+          .empty());
+
+  // Copy construction from a view constructor properties object with a label.
+  static_assert(std::is_copy_constructible_v<vcp_label_t>);
+
+  auto prop = Kokkos::view_alloc("our label");
+  vcp_label_t prop_copy(prop);
+
+  ASSERT_EQ(Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop),
+            "our label");
+  ASSERT_EQ(Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop_copy),
+            "our label");
```

Comment:
Also not quite convinced this is super useful.  This would pass with the current develop.

Reply:
The goal of this test is to ensure that the way this PR turns the constructor of the view constructor properties class into a forwarding constructor works well with the existing copy constructor. To ensure that it works well, we had to add a constraint to the forwarding constructor. We found that if we remove the constraint, the compiler will prefer the forwarding constructor over the copy constructor, and compilation of this test will fail.